src/Actions/BaseAction.php

```diff
@@ -73,9 +89,45 @@ public function getRules(): array
         return $this->rules;
     }
 
-    public function validate(): static
+    final public function execute()
+    {
+        if ($this->fireActionEvent(event: 'executing') === false) {
+            return false;
+        }
+
+        $returnValue = $this->performAction();
+
+        $this->fireActionEvent(event: 'executed', object: is_object($returnValue) ? $returnValue : null, halt: false);
+
+        return $returnValue;
+    }
+
+    final public function validate(): static
+    {
+        if ($this->fireActionEvent(event: 'validating') !== false) {
+            $this->validateData();
+
+            $this->fireActionEvent(event: 'validated', halt: false);
+        }
+
+        return $this;
+    }
+
+    protected function validateData(): void
     {
         $this->data = Validator::validate($this->data, $this->rules);
+    }
+
+    final public function withEvents(): static
```

Comment:
this is to enable events if previously disabled?

Reply:
yes, for example if you want to disable events on validating but enable them again on execute